### ISSUE TITLE
Create script to attach additional commit information to deployment

### ISSUE THREAD

**Issue by dinagraves** [GoogleCloudPlatform/fourkeys#52]
- Write script
- Update documentation

reference:
- https://github.com/GoogleCloudPlatform/fourkeys/pull/50
- https://github.com/GoogleCloudPlatform/fourkeys/issues/46

**Comment by rogerthatdev** [comment]
Hi @madeline-scyphers 

Here's the issue where we can track the work associated with our discussion today!

**Comment by madeline-scyphers** [comment]
I was thinking about how we talked about to do the data enrichment and make it its own `event_type` and I had a different idea I wanted to propose, a second table in addition to `events_raw` that is for just processed/enriched data with a FK back to `events_raw`. And then on any view that might use the processed data, you could just join the processed table and the raw table and take the processed data and then take the raw table in other spots if need be as a backup or for whatever other purpose.

**Comment by dinagraves** [comment]
Makes sense to me!

**Comment by madeline-scyphers** [comment]
Since this is gonna involve a few different changes, some structural like making a new table and a convenience view, and then others that will be the pub/sub stuff. Would it be helpful to break this up into multiple PRs either behind a feature flag, on a feature branch, or however, or I can just do it all at once. I just know larger reviews can be a bit of a pain, so I want to do whatever y'all think would be easier and best for y'all.

**Comment by dinagraves** [comment]
My vote is to do it in several PR's.  Don't worry about a feature flag or feature branch -- we can commit it to main, because it's a separate node that shouldn't interfere with anything currently running.  The last step we'll take is to add it to the terraform so people will pull it all together when it's ready.  I know you said you didn't have TF experience, so we can help with that part!  

Thank you again!

**Comment by madeline-scyphers** [comment]
A note about something I discovered, with the webhooks, at least with Gitlab (maybe with GitHub too), if a branch branches off of main and never comes back, the point at which it branch off of main is unknowable beyond for history. You can only get this history before if it comes back. The reason for this is because the webhooks don't treat this the way that default git does (but just having all commits be related with parents to get their history). But the webhooks do branches weirdly, the first commit in a branch off of main has the `"before"` attribute be the null sha, meaning no commit. 

```
{"object_kind": "push", 
"event_name": "push", 
"before": "0000000000000000000000000000000000000000", 
"after": "73f2658b6da76512b7d8b8a2ff4fc751737e8153",
...
```
How you get the history back is the commit that merges it back includes the `"commit"` attribute, which is a list of commits associated with that merge, 


```
"commits": [
    {"id": "954dcf5ee47025aafbf8e64cbbcaf97e08824670", ...}, 
    {"id": "58b6637656f3119fb7d7482e390d724e3de39089", ...}, 
    {"id": "73f2658b6da76512b7d8b8a2ff4fc751737e8153", ...}
]
```

But if you don't get a merge back, you never get that history back. 

So basically this will let people reconstruct what commits go with what releases as long as their branches get merged back into main, which considering fourkeys is assuming trunk based development, I think that is a fine assumptions. But I just wanted to mention it.

**Comment by madeline-scyphers** [comment]
Here is a checklist of things to get done (please reply to mention to add more)

TODO:

- [x] Add `events_enriched` table to store enriched data
- [x] add supporting infrastructure around `events_enriched` table (view connecting it to `events_raw`, add it to `refresh_data.sh`, add to setup directory, add insertion functions inside shared directory
- [ ] add to terraform directory modules
- [ ] Modify `generate_data.py` to generate "before" attribute so it can be used to test new algorithm properly
  - Add tests for new `generate_data.py` features
- [ ] Write algorithm to associate specific commits with a release and insert those commits into the `events_enriched` 
  - Should this be its own bq_worker or attached it as an additional step to the gitlab and github endpoint? If the later, then it can rely on those events already happening. If the former, is there a way we can ensure the associated main event has already happened? Also it needs the signature of the main event as a FK, would it be simpler to do something like this in `index` (new stuff at bottom)
  ```python
          if "headers" in attr:
              headers = json.loads(attr["headers"])
  
              # Process Gitlab Events
              if "X-Gitlab-Event" in headers:
                  event = process_gitlab_event(headers, msg)
  
          shared.insert_row_into_bigquery(event)
  
          if event:  # new stuff here
              enriched_event = process_enriched_event(event)
              shared.insert_row_into_events_enriched(enriched_event)
  ```
  - This would allow you to easily use the processed event without jumping through the hoops of processing it again or   calculating the signature again and then querying the db. @dinagraves thoughts?
- [ ] Update  `deployments.sql` to include `enriched_metadata` field if it is not null from the `events_enriched` table
- [ ] Update documentation to include new schema, any update information, etc.

potential addition
- [ ] Add script/sql query that inserts `events_enriched` row for previous deployments before this change
